### ISSUE TITLE
[PPO] Nan values with entropy regularization and action masking

### ISSUE THREAD

**Issue by Fabien-Couthouis** [tensorflow/agents#562]
Hello,

I am using PPOClipAgent and I have a problem with the entropy_regularization and action masking. I use a MaskSplitterNetwork to mask my actions:

```python
actor_wrapped_net = actor_distribution_network.ActorDistributionNetwork(observation_spec['observation'], action_spec, 
  fc_layer_params=PARAMS.actor_fc_layers, preprocessing_layers=preprocessing_layers, 
  preprocessing_combiner=preprocessing_combiner)

actor_net = mask_splitter_network.MaskSplitterNetwork(splitter_fn, actor_wrapped_net, passthrough_mask=True)

value_wrapped_net = value_network.ValueNetwork(
        observation_spec['observation'],
        fc_layer_params=PARAMS.value_fc_layers, preprocessing_layers=preprocessing_layers, preprocessing_combiner=preprocessing_combiner)

value_net = mask_splitter_network.MaskSplitterNetwork(
        splitter_fn, value_wrapped_net)
```

 In fact, when I set the entropy_regularization parameter in PPOClipAgent to a value > 0.0, I have an error coming from nan values found in  `entropy_reg_loss = tf.debugging.check_numerics(entropy_reg_loss, 'entropy_reg_loss')` in _ppo_agent.py_  (_entropy_regularization_loss_ method). I have set check_numerics=True in PPOClipAgent  params.

I dug into the error, it comes from this line: `per_example_loss = tf.math.multiply(per_example_loss, sample_weight)` in _common.py_ (_aggregate_losses_ function). As I mask my actions, some values of  `per_example_loss` are `inf` and even if my mask set the corresponding `sample_weight ` to 0, the multiplication results in a nan and so a nan entropy loss and a nan total loss.

I was able to launch the same code without nan and with the entropy regularization by replacing:

```python
per_example_loss = tf.math.multiply(per_example_loss, sample_weight)
```

by:

```python
per_example_loss = tf.math.multiply_no_nan(per_example_loss, sample_weight)
```
 in _common.py_ (_aggregate_losses_ function), but I hope there is a better solution.

Does anyone have an idea please?

Versions:

- OS: Windows 10 version 2004 (build 19041.746)
- Python 3.8.5
- tf_agents: 0.7.1
- tensorflow: 2.4.1
- tensorflow-probability: 0.12.1
- numpy: 1.20.0

**Comment by ebrevdo** [comment]
Thanks for reporting this!  Can you say a bit about why `per_example_loss` is coming out as `inf`?  I guess you are emitting values that are supposed to be zero-probability?  If you can debug that, it would help us.  In the meantime, I think `multiply_no_nan` makes sense - a discount of 0 means ignore that term.  We'll push a fix.

**Comment by ebrevdo** [comment]
(typically if you have `sample_weight` of 0 it means you're at an episode boundary or similar... is it possible at this locatoin, you're being fed a mask of all 0s, and this is what leads to the infs?)
